cli/operator/node.go

```diff
@@ -159,36 +189,100 @@ var StartNodeCmd = &cobra.Command{
 		}
 
 		cfg.SSVOptions.ValidatorOptions.DutyRoles = []spectypes.BeaconRole{spectypes.BNRoleAttester} // TODO could be better to set in other place
+
+		storageRoles := []spectypes.BeaconRole{
+			spectypes.BNRoleAttester,
+			spectypes.BNRoleProposer,
+			spectypes.BNRoleAggregator,
+			spectypes.BNRoleSyncCommittee,
+			spectypes.BNRoleSyncCommitteeContribution,
+			spectypes.BNRoleValidatorRegistration,
+		}
+		storageMap := ibftstorage.NewStores()
+
+		for _, storageRole := range storageRoles {
+			storageMap.Add(storageRole, ibftstorage.New(cfg.SSVOptions.ValidatorOptions.DB, storageRole.String(), cfg.SSVOptions.ValidatorOptions.ForkVersion))
+		}
+
+		cfg.SSVOptions.ValidatorOptions.StorageMap = storageMap
+		cfg.SSVOptions.ValidatorOptions.Metrics = metricsReporter
+
 		validatorCtrl := validator.NewController(logger, cfg.SSVOptions.ValidatorOptions)
 		cfg.SSVOptions.ValidatorController = validatorCtrl
+		cfg.SSVOptions.Metrics = metricsReporter
 
 		operatorNode = operator.New(logger, cfg.SSVOptions, slotTicker)
 
 		if cfg.MetricsAPIPort > 0 {
-			go startMetricsHandler(cmd.Context(), logger, db, cfg.MetricsAPIPort, cfg.EnableProfile)
+			go startMetricsHandler(cmd.Context(), logger, db, metricsReporter, cfg.MetricsAPIPort, cfg.EnableProfile)
 		}
 
-		nodeChecker.Wait()
+		nodeProber.Wait()
 
-		nodeUnreadyHandler := func() {
+		nodeProber.SetUnreadyHandler(func() {
 			logger.Fatal("Ethereum node(s) are either out of sync or down. Ensure the nodes are ready to resume.")
+		})
+
+		metricsReporter.SSVNodeHealthy()
+
+		eventDB := eventdb.NewEventDB(db.Badger())
+		eventDataHandler, err := eventdatahandler.New(
+			eventDB,
+			executionClient,
+			validatorCtrl,
+			cfg.SSVOptions.ValidatorOptions.OperatorData,
+			cfg.SSVOptions.ValidatorOptions.ShareEncryptionKeyProvider,
+			cfg.SSVOptions.ValidatorOptions.KeyManager,
+			cfg.SSVOptions.ValidatorOptions.Beacon,
+			storageMap,
+			eventdatahandler.WithFullNode(),
+			eventdatahandler.WithLogger(logger),
+			eventdatahandler.WithMetrics(metricsReporter),
+		)
+		if err != nil {
+			logger.Fatal("failed to setup event data handler", zap.Error(err))
+		}
+
+		eventBatcher := eventbatcher.NewEventBatcher()
+		eventDispatcher := eventdispatcher.New(
+			executionClient,
+			eventBatcher,
+			eventDataHandler,
+			eventdispatcher.WithLogger(logger),
+			eventdispatcher.WithMetrics(metricsReporter),
+			eventdispatcher.WithNodeProber(nodeProber),
+		)
+
+		txn := eventDB.ROTxn()
+		defer txn.Discard()
+
```

Comment:
why do we need it here? if thats the only way to access `GetLastProcessedBlock`, why not call `txn.Discard()` after getting it ?

Reply:
We need `txn.Discard()` because we must call it when we finish working with the transaction. Actually I could add a `GetLastProcessedBlock` method to event DB, but I think that would make event DB API less clean.

I agree that we can call `txn.Discard()` earlier. Fixed.